src/Extract.hs

```diff
@@ -181,73 +182,71 @@ metaType AbiStringType       = ByteStr
+idFromUpdate (BoolUpdate  item _) = idFromItem item
+idFromUpdate (BytesUpdate item _) = idFromItem item
+
+idFromLocation :: StorageLocation -> Id
+idFromLocation (IntLoc   item) = idFromItem item
+idFromLocation (BoolLoc  item) = idFromItem item
+idFromLocation (BytesLoc item) = idFromItem item
+
+contractFromRewrite :: Either StorageLocation StorageUpdate -> Id
+contractFromRewrite = either contractFromLoc contractFromUpdate
+
+contractFromItem :: TStorageItem t a -> Id
+contractFromItem (IntItem c _ _) = c
+contractFromItem (BoolItem c _ _) = c
+contractFromItem (BytesItem c _ _) = c
+
+ixsFromItem :: TStorageItem t a -> [TypedExp t]
+ixsFromItem (IntItem   _ _ ixs) = ixs
+ixsFromItem (BoolItem  _ _ ixs) = ixs
+ixsFromItem (BytesItem _ _ ixs) = ixs
 
 contractsInvolved :: Behaviour -> [Id]
-contractsInvolved beh =
-  getContractId . getLoc <$> _stateUpdates beh
-
-getContractId :: StorageLocation -> Id
-getContractId (IntLoc (DirectInt a _)) = a
-getContractId (BoolLoc (DirectBool a _)) = a
-getContractId (BytesLoc (DirectBytes a _)) = a
-getContractId (IntLoc (MappedInt a _ _)) = a
-getContractId (BoolLoc (MappedBool a _ _)) = a
-getContractId (BytesLoc (MappedBytes a _ _)) = a
-
-getContainerId :: StorageLocation -> Id
-getContainerId (IntLoc (DirectInt _ a)) = a
-getContainerId (BoolLoc (DirectBool _ a)) = a
-getContainerId (BytesLoc (DirectBytes _ a)) = a
-getContainerId (IntLoc (MappedInt _ a _)) = a
-getContainerId (BoolLoc (MappedBool _ a _)) = a
-getContainerId (BytesLoc (MappedBytes _ a _)) = a
-
-getContainerIxs :: StorageLocation -> [ReturnExp]
-getContainerIxs (IntLoc (DirectInt _ _)) = []
-getContainerIxs (BoolLoc (DirectBool _ _)) = []
-getContainerIxs (BytesLoc (DirectBytes _ _)) = []
-getContainerIxs (IntLoc (MappedInt _ _ ixs)) = NonEmpty.toList ixs
-getContainerIxs (BoolLoc (MappedBool _ _ ixs)) = NonEmpty.toList ixs
-getContainerIxs (BytesLoc (MappedBytes _ _ ixs)) = NonEmpty.toList ixs
+contractsInvolved = fmap contractFromRewrite . _stateUpdates
+
+contractFromLoc :: StorageLocation -> Id
+contractFromLoc (IntLoc item) = contractFromItem item
+contractFromLoc (BoolLoc item) = contractFromItem item
+contractFromLoc (BytesLoc item) = contractFromItem item
+
+contractFromUpdate :: StorageUpdate -> Id
+contractFromUpdate (IntUpdate item _) = contractFromItem item
+contractFromUpdate (BoolUpdate item _) = contractFromItem item
+contractFromUpdate (BytesUpdate item _) = contractFromItem item
+
+ixsFromLocation :: StorageLocation -> [TypedExp Untimed]
+ixsFromLocation (IntLoc item) = ixsFromItem item
+ixsFromLocation (BoolLoc item) = ixsFromItem item
+ixsFromLocation (BytesLoc item) = ixsFromItem item
+
+ixsFromUpdate :: StorageUpdate -> [TypedExp Untimed]
+ixsFromUpdate (IntUpdate item _) = ixsFromItem item
+ixsFromUpdate (BoolUpdate item _) = ixsFromItem item
+ixsFromUpdate (BytesUpdate item _) = ixsFromItem item
+
+ixsFromRewrite :: Either StorageLocation StorageUpdate -> [TypedExp Untimed]
```

Comment:
Yes, good call tbh. Is `Rewrite` a good name though? I struggled a bit coming up with anything nice. Like, `Rewrite` and `Update` are more or less synonyms so it's not really clear from the names which one is which. I was thinking `StorageEntry` or just `Entry` but that's kind of taken already. Of course, we could rename `TEntry` to `Storage`, `Read`, `Reference` or some combination thereof.

Reply:
Yeah I also can't really think of a name that feels totally satisfying. I think `Rewrite` is probably my favourite since if you squint a bit a `StorageLocation` is kind of just a rewrite where the poststate is the same as the prestate.

Reply:
Yeah, that was what I was thinking when I picked it too. Let's go with that for now, I just pushed a change.